### ISSUE TITLE
Training very slow, not fully utilizing CPU/GPU resources

### ISSUE THREAD

**Issue by ztqwerty** [Rikorose/DeepFilterNet#274]
Hi, Rikorose,

Great work on DeepFilterNet series! I have a question about the typical training time and CPU/GPU resources to be occupied:

I've generated the train/noise (~800G/60G) hdf5 files from DNS4 dataset, and attempt to use RTX3090 24G plus 20core CPU workstation to train by following the README instructions, and it is very slow. I modified the config.ini a little bit after it was generated, the current settings I am using are:

[train]
seed = 42
device =
model = deepfilternet2
mask_only = False
df_only = False
jit = False
batch_size = 64
batch_size_eval = 0
overfit = False
num_workers = 16
max_sample_len_s = 5.0
num_prefetch_batches = 32
global_ds_sampling_f = 1.0
dataloader_snrs = -5,0,5,10,20,40
batch_size_scheduling = 0/8,1/16,2/24,5/32,10/64,20/128,40/256
max_epochs = 60
validation_criteria = loss
validation_criteria_rule = min
early_stopping_patience = 5
start_eval = False

For epoch 0 there is >100k iterations scheduled, with each 100 iteration taking ~ 5 mins to run. However, seems like it is not taking all GPU/CPU resources, only taking ~ 3GB gpu memory, and barely any CPU resources. When checking with nvidia-smi command, the volatile GPU utilized is usually < 20%, sometimes 0%. Is this normal?

What hardware / config settings do you use for training? And typically how long does it take? Can you please share some tips on how to accelerate the training?

Thanks,
Tony

**Comment by Rikorose** [comment]
Hi,

I can't really reproduce your results. I need about 30-40sec for 100 iterations. I can utilize a quadro rtx 8000 pretty good.
One thing that you could try is:
```
 batch_size_scheduling = 0/32,4/64,16/128,32/256
```

**Comment by VanesYg** [comment]
Hi,I met the same problem

**Comment by pallaswept** [comment]
Hi there, sorry to bump this issue after so long, but I am seeing this same behaviour, with the same hardware. I have tried the above suggestion but there is no change. 

I see some power draw, about 112W (the card limit is 350W), but utilisation comes as momentary spikes and is mostly near 0%. VRAM utilisation is only 3672MB on this 24GB card.

@ztqwerty I wonder if you had any success with this? Perhaps you remember what was the trick? I'm really stuck, I hope you can help.

**Comment by pallaswept** [comment]
Sorry to have troubled you, I think this is resolved for me.

In a moment of desperation, I did the stupidest things: asked a search engine's AI, and tried to throw money at the problem - All general advice for slow training which I saw, suggested IO bottlenecking as the most likely cause, and my IO was very slow. I bought a new NVMe drive large enough for the entire dataset. Problem solved. Sometimes stupid still works, I guess :laughing: 

My 3090 sits on about 80% utilisation drawing 200W, completes 100 iterations of the training from the checkpoint in ~20 seconds (down from 10 minutes), 100 iterations of the first epoch of the LL model in 2-3 seconds (down from ~50 seconds). 

```
...
2024-10-22 11:44:00 | INFO     | DF | [120] [ 800/1213] | loss:  1.29482 | t_sample:  4.886E-05 | t_batch:  0.04273
2024-10-22 11:44:18 | INFO     | DF | [120] [ 900/1213] | loss:  1.29177 | t_sample:  5.575E-05 | t_batch:  0.03141
2024-10-22 11:44:37 | INFO     | DF | [120] [1000/1213] | loss:  1.29192 | t_sample:  2.946E-05 | t_batch:  0.02896
2024-10-22 11:44:56 | INFO     | DF | [120] [1100/1213] | loss:  1.28429 | t_sample:  4.912E-05 | t_batch:  0.03089
2024-10-22 11:45:14 | INFO     | DF | [120] [1200/1213] | loss:  1.26931 | t_sample:  1.906E-05 | t_batch:  0.02452
2024-10-22 11:45:16 | INFO     | DF | Avg batch loading time (py): 0.0002s
2024-10-22 11:45:16 | INFO     | DF | Avg batch loading time (rs): 0.0390s
2024-10-22 11:45:16 | INFO     | DF | Min/Avg/Max sample loading time (rs): 0.0000/0.0000/0.0438s
...
```

```
2024-10-22 12:08:51 | INFO     | DF | Start train epoch 0 with batch size 1
2024-10-22 12:08:52 | INFO     | DF | [0] [      0/1037280] | loss:  0.02722 | lr:  1.000E-04
2024-10-22 12:08:54 | INFO     | DF | [0] [    100/1037280] | loss:  0.13853 | lr:  1.000E-04
2024-10-22 12:08:57 | INFO     | DF | [0] [    200/1037280] | loss:  0.12079 | lr:  1.000E-04
2024-10-22 12:08:59 | INFO     | DF | [0] [    300/1037280] | loss:  0.11065 | lr:  1.000E-04
2024-10-22 12:09:01 | INFO     | DF | [0] [    400/1037280] | loss:  0.09797 | lr:  1.001E-04
2024-10-22 12:09:03 | INFO     | DF | [0] [    500/1037280] | loss:  0.07367 | lr:  1.001E-04
2024-10-22 12:09:06 | INFO     | DF | [0] [    600/1037280] | loss:  0.08249 | lr:  1.001E-04
...
```

Meanwhile, VRAM utilisation is still very low - even lower than before. So, I guess that behaviour is not symptomatic but is normal.
```
Wed Oct 23 00:21:01 2024
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 560.35.03              Driver Version: 560.35.03      CUDA Version: 12.6     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 3090        Off |   00000000:07:00.0  On |                  N/A |
| 65%   67C    P2            201W /  350W |    2104MiB /  24576MiB |     80%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
```

I hope this information is helpful to someone else in the future.
